compiler/rustc_middle/src/infer/canonical.rs

```diff
@@ -82,15 +82,32 @@ impl CanonicalVarValues<'_> {
     }
 
     pub fn is_identity_modulo_regions(&self) -> bool {
-        self.var_values.iter().enumerate().all(|(bv, arg)| match arg.unpack() {
-            ty::GenericArgKind::Lifetime(_) => true,
-            ty::GenericArgKind::Type(ty) => {
-                matches!(*ty.kind(), ty::Bound(ty::INNERMOST, bt) if bt.var.as_usize() == bv)
-            }
-            ty::GenericArgKind::Const(ct) => {
-                matches!(ct.kind(), ty::ConstKind::Bound(ty::INNERMOST, bc) if bc.as_usize() == bv)
+        let mut var = ty::BoundVar::from_u32(0);
+        for arg in self.var_values {
+            match arg.unpack() {
+                ty::GenericArgKind::Lifetime(_) => {},
+                ty::GenericArgKind::Type(ty) => {
```

Comment:
this code is subtle and feels a bit scary to me.

afaict what we want is:
- if this is the first non-region argument, check that its a bound var with index `n`
- for all following arguments, check that they are bound vars whose index each increases by 1, starting from `n`

While this is guaranteed by your impl afaict, I think it might be cleaner to have `let mut expected_non_region_bv = None` and then have `expected_non_region_bv.map_or(true, |expected| bt.var == expected)` ... `expected_non_region_bv = Some(bt.var + 1);`